### ISSUE TITLE
Eloquent returns empty collection same table

### ISSUE THREAD

**Issue by ceesvanegmond** [laravel/framework#4754]
Consider the following:
```php
class Post extends Eloquent
{
    protected $table = 'posts';

    public function children()
    {
        return $this->hasMany('Post', 'parent_id')
    }

    public function parent()
    {
        return $this->belongsTo('Post', 'parent_id');
    }
}
```

So, i'm having a `tree like` configuration. When I have an children post and I try to get the parent of this children post, the following SQL query is returned from Eloquent.
```sql
select * from `posts` where (select count(*) from `posts` where `posts`.`parent_id` = `posts`.`id`)
```

which returns an empty ```Collection```. When debugging, I found out that the ```Collection``` is empty because of the use the word ```posts``` in the query. What would work is to alias those tables, like:
```sql
select * from posts u where (select count(*) from posts p where p.parent_id = u.id)
```

Is there something with `Eloquent` we can use to solve this, or this has to be an custom query?

**Comment by GrahamCampbell** [comment]
> Isn't that a pretty common use case?

Can't be if it doesn't work? Is anyone able to look into this please? If not, this may end up getting branded as a "nofix".

**Comment by franzliedke** [comment]
This [has been fixed before](https://github.com/laravel/framework/commit/16f099e0bc17d89f1e11aff6343b2999230560a3) for `belongsToMany` relationships.

I'm playing around with it, but not sure whether I can come up with a fix. @taylorotwell may be able to fix this more easily.

**Comment by franzliedke** [comment]
Can be closed - this was recently fixed in #8193.
